### PR TITLE
Save Controls

### PR DESCRIPTION
**Pull request by a-crowell**
### Description
Make control schemes save between runs of Synthesis.

### Objectives
- [X] Save robot control schemes
- [X] Save global controls
- [X] Session Save vs Save vs Cancel option
- [X] Robot tooltips shows accurate keybinds instead of default
- [X] Edited controls are highlighted to make clearer which will be saved

### Testing Done
- Changing and checking if they save between runs: different drivebases, all of Team 2471 v7's joints, and intake/shoot/god mode
- Ensuring changes don't save if the Session button is clicked instead of Save
- Robot tooltips is accurate through those tests

[JIRA Issue](https://jira.autodesk.com/browse/AARD-1584)

### CODE REVIEW

**Comment by a-crowell** [comment]
> Nice job! I'm not sure if this was intended, but when you have multiple different robots, they are both affected by the normal and session save.
> 
> The only other issue I found is that when you spawn two of the same robots, remove one, then edit controls you get this exception: image

Fixed the multiple bots getting affected, but less sure about the fix for the error. Let me know if it shows up again.

**Comment by a-crowell** [comment]
Don't pull right now, breaks other branches.

**Comment by a-crowell** [comment]
> Don't pull right now, breaks other branches.

Wait until Save Controls Compatibility is in or uninstall and reinstall your mirabuf files when you're done to avoid errors when you switch branches.

**Review by Markytt — CHANGES_REQUESTED** [review]
Everything is working well, I didn't find any new bugs. Right now though, there is no way to close the controls panel without saving or doing a sessions save, so it would be nice if a close or cancel button was added to the controls panel.

File: engine/Assets/Scripts/Behaviors/SwerveDriveBehaviour.cs [hunk 1]
```diff
@@ -12,6 +13,14 @@
 
 namespace Synthesis {
     public class SwerveDriveBehaviour : SimBehaviour {
+        private string forward             = "Swerve Forward";
+        private string backward            = "Swerve Backward";
+        private string left                = "Swerve Left";
+        private string right               = "Swerve Right";
+        private string turn_left           = "Swerve Turn Left";
+        private string turn_right          = "Swerve Turn Right";
+        private string reset_field_forward = "Swerve Reset Forward";
```
Comment (KyroVibe) [inline engine/Assets/Scripts/Behaviors/SwerveDriveBehaviour.cs:22]: You forgot to append the id to the strings like how you did in ArcadeDrive

File: engine/Assets/Scripts/Behaviors/TankDriveBehavior.cs [hunk 2]
```diff
@@ -12,6 +13,11 @@
 
 namespace Synthesis {
     public class TankDriveBehavior : SimBehaviour {
+        private string left_forward  = "Tank Left-Forward";
+        private string left_reverse  = "Tank Left-Reverse";
+        private string right_forward = "Tank Right-Forward";
+        private string right_reverse = "Tank Right-Reverse";
```
Comment (KyroVibe) [inline engine/Assets/Scripts/Behaviors/TankDriveBehavior.cs:19]: Same issue as in SwerveDrive

File: engine/Assets/Scripts/Behaviors/ArcadeDriveBehaviour.cs [hunk 3]
```diff
@@ -50,17 +55,22 @@ public class ArcadeDriveBehaviour : SimBehaviour {
             _leftWheels  = leftWheels;
             _rightWheels = rightWheels;
 
+            forward  = SimObjectId + FORWARD;
```
Comment (KyroVibe) [inline engine/Assets/Scripts/Behaviors/ArcadeDriveBehaviour.cs:58]: This needs to be `MiraLive.MiraAssembly.Info.GUID` so it saves for multiple iterations of the same bot

**Review by KyroVibe — CHANGES_REQUESTED** [review]
Save and Session saved worked for robot controls. There are couple of things that need to be fixed however.

**Review by Markytt — COMMENTED** [review]
The controls sometimes aren't saving when you change robots, or when you exit and re-enter practice mode. Also when you stop the program and run it again, the controls get reset to what they originally were.

**Comment by a-crowell** [comment]
> The controls sometimes aren't saving when you change robots, or when you exit and re-enter practice mode. Also when you stop the program and run it again, the controls get reset to what they originally were.

Are you clicking Save, not Session Save? If so, may I ask which robots/controls you're changing?

**Review by LucaHaverty — CHANGES_REQUESTED** [review]
I found a few issues:

- Global control changes persist even if you click session save
- If you edit a robot control, click session, then enter the control modal again and click save, the changes still reset if you stop and play unity
- If you modify a global control and save without a field or robot spawned, you get
image

- If I edit controls mean machine then save or session save, there is a bit lag spike (multiple seconds). I haven't seen this happen with dozer

Saving and session saving robot controls seems to work well, nice work!

**Comment by a-crowell** [comment]
> I found a few issues:
> 
> * Global control changes persist even if you click session save
> * If you edit a robot control, click session, then enter the control modal again and click save, the changes still reset if you stop and play unity
> * If you modify a global control and save without a field or robot spawned, you get
>   image
> * If I edit controls mean machine then save or session save, there is a bit lag spike (multiple seconds). I haven't seen this happen with dozer
> 
> Saving and session saving robot controls seems to work well, nice work!

Will get to work on the global control issues and look into optimizing Mean Machine! I figured we wouldn't want controls edited from the last time users opened the panel to Save, just so you can Session Save vs Save specific controls. What do y'all prefer? For now, I'll highlight the edited controls to make that clearer. Do y'all think something would be useful on the buttons themselves? Like "Save (2)" for 2 controls changed?

**Comment by LucaHaverty** [comment]
> > I found a few issues:
> > 
> > * Global control changes persist even if you click session save
> > * If you edit a robot control, click session, then enter the control modal again and click save, the changes still reset if you stop and play unity
> > * If you modify a global control and save without a field or robot spawned, you get
> >   image
> > * If I edit controls mean machine then save or session save, there is a bit lag spike (multiple seconds). I haven't seen this happen with dozer
> > 
> > Saving and session saving robot controls seems to work well, nice work!
> 
> Will get to work on the global control issues and look into optimizing Mean Machine! I figured we wouldn't want controls edited from the last time users opened the panel to Save, just so you can Session Save vs Save specific controls. What do y'all prefer? For now, I'll highlight the edited controls to make that clearer. Do y'all think something would be useful on the buttons themselves? Like "Save (2)" for 2 controls changed?

Yeah I think that's a good idea

**Review by KyroVibe — CHANGES_REQUESTED** [review]
So very very close. When I session save, then spawn another one of the same bot, the controls reset to what they were before I session saved

**Review by LucaHaverty — APPROVED** [review]
I didn't find any issues. I like how changed controls are highlighted, great work!

File: engine/Assets/Scripts/Behaviors/ArcadeDriveBehaviour.cs [hunk 4]
```diff
@@ -38,58 +43,71 @@ public class ArcadeDriveBehaviour : SimBehaviour {
...
-                    break;
+            if (args.InputKey.Length > MiraId.Length) {
+                string s = args.InputKey.Remove(0, MiraId.Length);
+                switch (s) {
+                    case FORWARD:
+                    case BACKWARD:
+                    case LEFT:
+                    case RIGHT:
+                        if (base.MiraId != MainHUD.SelectedRobot.MiraGUID ||
+                            !(DynamicUIManager.ActiveModal as ChangeInputsModal).isSave)
```
Comment (KyroVibe) [inline engine/Assets/Scripts/Behaviors/ArcadeDriveBehaviour.cs:98]:
```suggestion
                            (DynamicUIManager.ActiveModal as ChangeInputsModal)?.isSave ?? false)
```

**Review by KyroVibe — CHANGES_REQUESTED** [review]
Awesome work, add one more null check and you're done

**Review by KyroVibe — APPROVED** [review]
Great work, just added a quick null check

**Review by KyroVibe — APPROVED** [review]

**Pull request merged**